csrc/multidevice/executor.cpp

```diff
@@ -6,133 +6,234 @@
+      if (!tv->axis(static_cast<int>(i))->isDeviceDim()) {
+        tv->axis(static_cast<int>(i))->parallelize(ParallelType::Serial);
+      }
+    }
+  }
+
+  return std::
+      make_pair<std::unique_ptr<Fusion>, std::unordered_map<Val*, Val*>>(
+          std::move(fusion_copy), std::move(copy_to_original_map));
+}
+
+} // namespace
+
+// TODO: use native allocator instead.
+std::unordered_map<Val*, c10::IValue> MultiDeviceExecutor::allocateRecvBuffers(
+    std::vector<c10::IValue> global_inputs_IValues) {
+  std::unordered_set<Val*> vals_to_allocate;
+  std::unordered_set<Val*> vals_to_not_allocate;
+  for (auto group : staged_fusion_->groups()) {
+    if (is_resharding_[group]) {
+      NVF_ERROR(group->exprs().size() == 1);
+      NVF_ERROR(group->exprs().at(0)->outputs().size() == 1);
+      auto val = group->exprs().at(0)->outputs().at(0);
+      NVF_ERROR(val->isA<TensorView>());
+      auto tv = val->as<TensorView>();
+      NVF_ERROR(tv->hasDeviceMesh());
+      if (tv->getDeviceMesh().has(comm_.deviceId())) {
+        vals_to_allocate.insert(val);
+      }
+    }
+  }
+
+  auto [fusion_copy, copy_to_original_map] =
+      copyFusionAndChangeOutputs(fusion(), vals_to_allocate);
+  if (fusion_copy->outputs().empty()) {
+    return {};
+  }
+  // TODO: Not working with FusionExecutorCache
+  FusionExecutor fe;
+  fe.compileFusion(fusion_copy.get(), global_inputs_IValues);
+  auto buffers = fe.allocOutputSpace(global_inputs_IValues);
+
+  std::unordered_map<Val*, c10::IValue> allocations;
+  for (auto i : c10::irange(buffers.size())) {
+    allocations.emplace(
+        copy_to_original_map[fusion_copy->outputs().at(i)], buffers.at(i));
+  }
+
+  return allocations;
+}
+
+MultiDeviceExecutor::MultiDeviceExecutor(
+    std::unique_ptr<Fusion> fusion,
+    Communicator& comm)
+    : comm_(comm) {
+  insertReshardings(fusion.get());
+  SegmentCandidateFinderOptions options{
+      .run_translate_welford = false,
+      .run_combine_reductions = false,
+      .run_herrmann_merge = true,
+      .run_final_merge = true,
+      .only_segment_resharding_exprs = true};
+
+  staged_fusion_ =
+      SegmentCandidateFinder::segment(std::move(fusion), nullptr, options);
+
+  for (auto group : staged_fusion_->groups()) {
+    NVF_ERROR(!group->exprs().empty() == 1, "invalid segmentation");
+    is_resharding_[group] = std::any_of(
+        group->exprs().begin(), group->exprs().end(), [](auto expr) {
+          return isResharding(expr);
+        });
+    NVF_ERROR(
+        !is_resharding_[group] || group->exprs().size() == 1,
+        "Communications cannot be fused");
+    auto expr = group->exprs().at(0);
+    should_run_[group] = involvedDevices(expr).count(comm_.deviceId());
   }
-  return should_run_[stage];
+  // prepare the order in which to launch the kernels/comms
+  RuntimeWorkSpace workspace;
+  prepareRuntimeOrder(staged_fusion_.get(), workspace);
+  group_run_order_ = std::move(workspace.group_run_order);
```

Comment:
I'd just have a `RuntimeWorkSpace` in `MultiDeviceExecutor` rather than just `group_run_order`. The other vector would be eventually required.

Reply:
no problem, I add it